File: Combination Sum - Leetcode 39/Combination Sum - Leetcode 39.java
import java.util.*;

public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> sol = new ArrayList<>();
        backtrack(candidates, target, 0, 0, sol, res);
        return res;
    }

    private void backtrack(int[] candidates, int target, int start, int curSum, List<Integer> sol, List<List<Integer>> res) {
        if (curSum == target) {
            res.add(new ArrayList<>(sol));
            return;
        }
        if (curSum > target || start == candidates.length) {
            return;
        }

        // Skip the current candidate
        backtrack(candidates, target, start + 1, curSum, sol, res);

        // Include the current candidate
        sol.add(candidates[start]);
        backtrack(candidates, target, start, curSum + candidates[start], sol, res);
        sol.remove(sol.size() - 1);
    }
}

File: Combination Sum - Leetcode 39/Combination Sum - Leetcode 39.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> sol;
        backtrack(candidates, target, 0, 0, sol, res);
        return res;
    }

private:
    void backtrack(vector<int>& candidates, int target, int start, int curSum, vector<int>& sol, vector<vector<int>>& res) {
        if (curSum == target) {
            res.push_back(sol);
            return;
        }
        if (curSum > target || start == candidates.size()) {
            return;
        }

        // Skip the current candidate
        backtrack(candidates, target, start + 1, curSum, sol, res);

        // Include the current candidate
        sol.push_back(candidates[start]);
        backtrack(candidates, target, start, curSum + candidates[start], sol, res);
        sol.pop_back();
    }
};
